lua/nvim-treesitter/textobjects/move.lua

```diff
@@ -70,16 +74,158 @@ local function move(query_strings, forward, start, winid)
 end
 
 M.goto_next_start = function(query_strings)
```

Comment:
can't this be generalized? By transforming wrapping a function? This could be used to make external functions of the user repeatable. Like
``` lua
function M.make_repeatable_move(move_fn)
    return function(...)
         M.last_move = { func = move_fn, args = {...} }
         move_fn(...)
    end
end

Reply:
or even 
```
function M.make_repeatable_move(move_fn)
    return function(...)
         M.last_move = function() move_fn(...) end
         move_fn(...)
    end
end
```

Reply:
The only problem is that when you repeat movement you need to know the direction. Unfortunately, the argument position of `forward` is in a random place at the moment. So in `M.repeat_last_move_next` for example I had to recognise the name of the function and change the `forward` value to `true`.

Reply:
If we can move the forward argument position to always the first maybe we can use this design

Reply:
Or, just remove all arguments except accepting one opts table with `forward=` key.

Reply:
ah, I get it now. Because you can repeat but in opposite direction.

Reply:
so you would need to define `make_repeatable_move` in terms of a `forward_fn` and `backwarn_fn` pair.

Reply:
Yes, but I think making it extensible is a good idea. Other plugins can just make a wrapper function that follows the rule (opt should have forward key).

I also want to use it in my plugin.

Reply:
> so you would need to define `make_repeatable_move` in terms of a `forward_fn` and `backwarn_fn` pair.
> 
> We can leave it as is. So every function to extend from user space must have the direction as boolish first parameter to work.

Separating with two functions is also a good idea. Should we choose the table argument design though?(maybe more future proof as it can accept more than forward, like start and end etc. if it matters)

Reply:
Yes!!! I took that idea already even further: https://github.com/ghostbuster91/nvim-next

I don't have any experience in developing nvim plugins so any feedback appreciated. If you would like to help me with this effort just let me know and I would be happy to add you as maintainers :)

Reply:
yes, it should be fine as it is now. But it would be nice to have a function that would automatically turn a user's function into a repeatable one, so that they don't need to understand how `last_move` works. Docs can explain that the first parameter needs to be the direction.

Reply:
I like @ghostbuster91's approach

Reply:
never mind, we can make it accept both boolean and table in the future so it is still future proof. Implementing this now.

Reply:
I just pushed the change.

Reply:
@theHamsta I am glad to hear that. I don't think that it is possible to have this repeatable magic without any manual call (unless that is somehow implemented directly in the core). But again I am quite new to this space so I might be missing something obvious.

Reply:
> @theHamsta I am glad to hear that. I don't think that it is possible to have this repeatable magic without any manual call (unless that is somehow implemented directly in the core). But again I am quite new to this space so I might be missing something obvious.

I just saw your repo. Thanks for sharing!  
It would have been really nice if it was nvim's builtin to register repeatability, or plugins like plenary that everyone has.  
However, requiring more dependencies and further setup call isn't ideal.

The idea is that instead of requiring a plugin and setup call, you can make plugin developers use our function to make it repeatable, or making users turn existing plugins' movement into a repeatable one. It is possible with `make_repeatable_move` now.

Reply:
@kiyoon so your idea was to make the first arg a table to be more future proof?

I think there are the following approaches:
- make function magical by having it the first arg boolean
- make function magical by having it the first arg a table
- make function pair magical and use both return values to have two magical functions
- not have magical functions but set up everything via config

Reply:
I think the first arg can be a boolean for now, and if we want to extend it we can use a table and check the type of the first arg. So the approach 1 and 2 are kind of similar.

Reply:
We can provide option 3 along with the one we have. I'll make this as well.

Reply:
I just pushed the change.

Reply:
I think option 4 is limiting because other plugins can't use our functions under the hood, and users need to config everything right?

Reply:
I don't know if this is a general concern. If move_fn accepts buf_id or win_id, the repeat will only perform at the same buffer not the current buffer. Should we have a feature to automatically set them to 0?

Reply:
> I don't know if this is a general concern. If move_fn accepts buf_id or win_id, the repeat will only perform at the same buffer not the current buffer. Should we have a feature to automatically set them to 0?

But honestly users wouldn't mind using the move command again and then perform repeat. Who would move around the buffers and use repeat?

Reply:
> * make function magical by having it the first arg boolean
> * make function magical by having it the first arg a table

If we introduce the table stuff later, then whoever is directly accessing `M.last_move` to perform whatever custom stuff they want to do will face problems later.  
So we can reserve the table usage in the doc and implement it now, so it doesn't cause any confusion in the future.

Reply:
> The idea is that instead of requiring a plugin and setup call, you can make plugin developers use our function to make it repeatable, or making users turn existing plugins' movement into a repeatable one. It is possible with make_repeatable_move now.

Sure, the setup call is not necessary, it is just a convenient way of setting it. There is already an option to just use the wrapper method and turn any function into a repeatable pair, just as you did (I've updated the readme to make it clear). 

I think that it is better to have it as a separate plugin than as a part of nvim-treesitter-textobjects since not everyone needs to have your plugin installed in order to benefit from custom repeatable movements. IMO It also fits better into nvim philosophy where you have many small things that you compose together.

Reply:
Update: I just noticed that supporting both types is infeasible as we don't know how the custom function behaves.  
Should we use a boolean or a table?

Reply:
@ghostbuster91 That's true. I think there are many plugins that are supposedly doing this, but for some reason I can't get them to work nicely with lua functions.

Reply:
I think this plugin can offer a native solution to do this, so everyone using treesitter textobjects benefit from this functionality. However, it is completely optional whether you want to extend it to a separate plugin and utilise other plugins at the same time. This way, most people wouldn't have to change settings to use repeat motions, but some people can optionally do this step to make it more customised.

Reply:
@kiyoon you think `M.last_move` can be private now that users have some functions that doesn't need them understand `last_move`?

Reply:
> @kiyoon you think `M.last_move` can be private now that users have some functions that doesn't need them understand `last_move`?

There are some more complicated move behaviours where they want to change the arguments for the repeat functionality.
Builtin t etc. are the examples where you want the repeat functionality to be different from just executing the same function again.
So maybe leaving it open will make it more general to be used by others.

Wait I can edit your comment as well? I did this accidentally and I'm super confused

Reply:
@theHamsta Furthermore, people can even customise the last move bahaviour by making their custom function that uses `M.last_move`.

For example, they can make `repeat_last_move_next_start_inner`